File: Research/RouterTrace/routide_trace/process_memory.py
"""Validate sampled process memory separately from MLX allocator accounting."""

from __future__ import annotations

import math
from typing import Any

PROCESS_MEMORY_SCOPE = "generation-request-including-model-preparation-and-metric-drain"
PEAK_SCOPE = "maximum-observed-sample-within-request-not-a-continuous-high-water-mark"


def _number(value: Any, label: str) -> float:
    if type(value) not in (int, float) or not math.isfinite(value):
        raise ValueError(f"processMemory {label} must be finite")
    return float(value)


def _integer(value: Any, label: str) -> int:
    if type(value) is not int or not 0 <= value <= 2**64 - 1:
        raise ValueError(f"processMemory {label} must be a nonnegative integer")
    return value


def validate_process_memory(report: Any) -> None:
    if not isinstance(report, dict) or type(report.get("schemaVersion")) is not int:
        raise ValueError("processMemory must contain a versioned report")
    if (
        report["schemaVersion"] != 1
        or report.get("scope") != PROCESS_MEMORY_SCOPE
        or report.get("peakScope") != PEAK_SCOPE
        or report.get("method") != "task_info(TASK_VM_INFO): resident_size and phys_footprint"
        or report.get("availableMemoryMeaning") != "current-process dirty-memory-limit headroom; not system free RAM"
    ):
        raise ValueError("processMemory measurement contract is not recognized")
    for key in ("availableMemorySupported", "modelWasLoadedAtStart"):
        if type(report.get(key)) is not bool:
            raise ValueError(f"processMemory {key} must be a recorded Boolean")
    if _integer(report.get("sampleIntervalMilliseconds"), "sample interval") != 250:
        raise ValueError("processMemory requires the declared 250 ms sampling interval")
    _integer(report.get("lifecycleInterruptions"), "lifecycle interruptions")
    if report.get("memoryWarnings") is not None:
        _integer(report["memoryWarnings"], "memory warnings")
    for key in ("startedAtUnixSeconds", "finishedAtUnixSeconds"):
        if _number(report.get(key), key) <= 0:
            raise ValueError(f"processMemory {key} must be positive")
    duration = _number(report.get("monotonicDurationSeconds"), "duration")
    if duration < 0:
        raise ValueError("processMemory duration must be nonnegative")
    samples = report.get("samples")
    failures = report.get("samplingFailures")
    if not isinstance(samples, list) or not isinstance(failures, list):
        raise ValueError("processMemory requires samples and explicit samplingFailures")
    attempts = []
    for records, failed in ((samples, False), (failures, True)):
        previous = -1.0
        for item in records:
            if not isinstance(item, dict) or item.get("trigger") not in ("start", "periodic", "manual", "end"):
                raise ValueError("processMemory has an invalid sample trigger")
            elapsed = _number(item.get("elapsedSeconds"), "sample elapsedSeconds")
            if not previous <= elapsed <= duration or elapsed < 0:
                raise ValueError("processMemory samples must be ordered within the request")
            previous = elapsed
            attempts.append((elapsed, item["trigger"]))
            if failed:
                if not isinstance(item.get("message"), str) or not item["message"]:
                    raise ValueError("processMemory sampling failures need an explicit message")
                continue
            reading = item.get("reading")
            if not isinstance(reading, dict):
                raise ValueError("processMemory sample reading is missing")
            _integer(reading.get("residentBytes"), "RSS")
            _integer(reading.get("physicalFootprintBytes"), "physical footprint")
            if report["availableMemorySupported"]:
                _integer(reading.get("availableMemoryBytes"), "available process memory")
            elif reading.get("availableMemoryBytes") is not None:
                raise ValueError("processMemory reports unsupported available memory")
    attempts.sort()
    triggers = [trigger for _, trigger in attempts]
    if (
        triggers.count("start") != 1 or triggers.count("end") != 1
        or triggers[0] != "start" or triggers[-1] != "end"
    ):
        raise ValueError("processMemory must attempt both request-boundary samples exactly once")
    expected_status = "unavailable" if not samples else "partial" if failures else "complete"
    if report.get("samplingStatus") != expected_status:
        raise ValueError("processMemory samplingStatus conceals missing or failed samples")
    if _integer(report.get("sampleCount"), "sample count") != len(samples):
        raise ValueError("processMemory sampleCount does not match its samples")
    for key, trigger in (("baseline", "start"), ("final", "end")):
        expected = next((item["reading"] for item in samples if item["trigger"] == trigger), None)
        if report.get(key) != expected:
            raise ValueError(f"processMemory {key} does not match the boundary sample")
    for key, reading_key, reduce in (
        ("peakResidentBytes", "residentBytes", max),
        ("peakPhysicalFootprintBytes", "physicalFootprintBytes", max),
        ("minimumAvailableMemoryBytes", "availableMemoryBytes", min),
    ):
        values = [item["reading"][reading_key] for item in samples if item["reading"].get(reading_key) is not None]
        expected = reduce(values) if values else None
        observed = report.get(key)
        if observed is not None:
            _integer(observed, key)
        if observed != expected:
            raise ValueError(f"processMemory {key} does not reconcile with sampled readings")
    boundaries = [0.0] + [item["elapsedSeconds"] for item in samples] + [duration]
    expected_gap = max(right - left for left, right in zip(boundaries, boundaries[1:]))
    if not math.isclose(
        _number(report.get("maximumSamplingGapSeconds"), "maximum sampling gap"),
        expected_gap, rel_tol=1e-12, abs_tol=1e-9,
    ):
        raise ValueError("processMemory maximum sampling gap does not match successful samples")

File: Sources/RoutideRuntime/ProcessMemory.swift
import Darwin
import Foundation
import os

public struct ProcessMemoryReading: Codable, Equatable, Sendable {
    public let residentBytes: UInt64
    public let physicalFootprintBytes: UInt64
    public let availableMemoryBytes: UInt64?

    public init(
        residentBytes: UInt64, physicalFootprintBytes: UInt64, availableMemoryBytes: UInt64?
    ) {
        self.residentBytes = residentBytes
        self.physicalFootprintBytes = physicalFootprintBytes
        self.availableMemoryBytes = availableMemoryBytes
    }
}

public enum ProcessMemoryError: Error, LocalizedError, Sendable {
    case machFailure(Int32)
    case incompleteTaskInfo

    public var errorDescription: String? {
        switch self {
        case .machFailure(let code):
            "task_info(TASK_VM_INFO) failed with Mach error \(code)."
        case .incompleteTaskInfo:
            "TASK_VM_INFO did not return the physical-footprint field."
        }
    }
}

public enum SystemProcessMemory {
    public static var availableMemorySupported: Bool {
        #if os(iOS) && !targetEnvironment(macCatalyst)
            true
        #else
            false
        #endif
    }

    public static func read() -> Result<ProcessMemoryReading, ProcessMemoryError> {
        var info = task_vm_info_data_t()
        var count = mach_msg_type_number_t(
            MemoryLayout<task_vm_info_data_t>.stride / MemoryLayout<integer_t>.stride
        )
        let capacity = Int(count)
        let result = withUnsafeMutablePointer(to: &info) { pointer in
            pointer.withMemoryRebound(to: integer_t.self, capacity: capacity) {
                task_info(mach_task_self_, task_flavor_t(TASK_VM_INFO), $0, &count)
            }
        }
        guard result == KERN_SUCCESS else {
            return .failure(.machFailure(result))
        }
        guard let offset = MemoryLayout<task_vm_info_data_t>.offset(of: \.phys_footprint),
            Int(count) * MemoryLayout<integer_t>.stride >= offset + MemoryLayout<UInt64>.size
        else {
            return .failure(.incompleteTaskInfo)
        }
        let available: UInt64?
        #if os(iOS) && !targetEnvironment(macCatalyst)
            available = UInt64(os_proc_available_memory())
        #else
            available = nil
        #endif
        return .success(
            ProcessMemoryReading(
                residentBytes: info.resident_size,
                physicalFootprintBytes: info.phys_footprint,
                availableMemoryBytes: available
            ))
    }
}

public enum ProcessMemorySampleTrigger: String, Codable, Sendable {
    case start, periodic, manual, end
}

public struct ProcessMemorySample: Codable, Equatable, Sendable {
    public let trigger: ProcessMemorySampleTrigger
    public let elapsedSeconds: Double
    public let reading: ProcessMemoryReading
}

public struct ProcessMemorySampleFailure: Codable, Equatable, Sendable {
    public let trigger: ProcessMemorySampleTrigger
    public let elapsedSeconds: Double
    public let message: String
}

public struct ProcessMemoryReport: Encodable, Sendable {
    public let schemaVersion = 1
    public let scope = "generation-request-including-model-preparation-and-metric-drain"
    public let peakScope = "maximum-observed-sample-within-request-not-a-continuous-high-water-mark"
    public let method = "task_info(TASK_VM_INFO): resident_size and phys_footprint"
    public let availableMemoryMeaning =
        "current-process dirty-memory-limit headroom; not system free RAM"
    public let startedAtUnixSeconds: Double
    public let finishedAtUnixSeconds: Double
    public let monotonicDurationSeconds: Double
    public let sampleIntervalMilliseconds: Int
    public let availableMemorySupported: Bool
    public let modelWasLoadedAtStart: Bool
    public let memoryWarnings: Int?
    public let lifecycleInterruptions: Int
    public let samples: [ProcessMemorySample]
    public let samplingFailures: [ProcessMemorySampleFailure]
    public let samplingStatus: String
    public let sampleCount: Int
    public let baseline: ProcessMemoryReading?
    public let final: ProcessMemoryReading?
    public let peakResidentBytes: UInt64?
    public let peakPhysicalFootprintBytes: UInt64?
    public let minimumAvailableMemoryBytes: UInt64?
    public let maximumSamplingGapSeconds: Double

    init(
        startedAtUnixSeconds: Double,
        finishedAtUnixSeconds: Double,
        monotonicDurationSeconds: Double,
        sampleIntervalMilliseconds: Int,
        availableMemorySupported: Bool,
        modelWasLoadedAtStart: Bool,
        memoryWarnings: Int?,
        lifecycleInterruptions: Int,
        samples: [ProcessMemorySample],
        samplingFailures: [ProcessMemorySampleFailure]
    ) {
        self.startedAtUnixSeconds = startedAtUnixSeconds
        self.finishedAtUnixSeconds = finishedAtUnixSeconds
        self.monotonicDurationSeconds = monotonicDurationSeconds
        self.sampleIntervalMilliseconds = sampleIntervalMilliseconds
        self.availableMemorySupported = availableMemorySupported
        self.modelWasLoadedAtStart = modelWasLoadedAtStart
        self.memoryWarnings = memoryWarnings
        self.lifecycleInterruptions = lifecycleInterruptions
        self.samples = samples
        self.samplingFailures = samplingFailures
        self.samplingStatus =
            samples.isEmpty ? "unavailable" : samplingFailures.isEmpty ? "complete" : "partial"
        self.sampleCount = samples.count
        self.baseline = samples.first(where: { $0.trigger == .start })?.reading
        self.final = samples.last(where: { $0.trigger == .end })?.reading
        self.peakResidentBytes = samples.map(\.reading.residentBytes).max()
        self.peakPhysicalFootprintBytes = samples.map(\.reading.physicalFootprintBytes).max()
        self.minimumAvailableMemoryBytes = samples.compactMap(\.reading.availableMemoryBytes).min()
        let boundaries = [0.0] + samples.map(\.elapsedSeconds) + [monotonicDurationSeconds]
        self.maximumSamplingGapSeconds =
            zip(boundaries, boundaries.dropFirst())
            .map { max(0, $1 - $0) }.max() ?? monotonicDurationSeconds
    }
}

public final class ProcessMemorySampler: @unchecked Sendable {
    public static let intervalMilliseconds = 250
    public typealias Reader = @Sendable () -> Result<ProcessMemoryReading, ProcessMemoryError>

    private let queue = DispatchQueue(label: "research.Routide.process-memory")
    private let reader: Reader
    private let startClock = ContinuousClock.now
    private let startDate = Date()
    private let modelWasLoadedAtStart: Bool
    private let availableMemorySupported: Bool
    private let logger = Logger(subsystem: "research.Routide", category: "ProcessMemory")
    private var timer: DispatchSourceTimer?
    private var samples: [ProcessMemorySample] = []
    private var failures: [ProcessMemorySampleFailure] = []
    private var finishedReport: ProcessMemoryReport?

    public init(
        modelWasLoadedAtStart: Bool,
        availableMemorySupported: Bool = SystemProcessMemory.availableMemorySupported,
        reader: @escaping Reader = { SystemProcessMemory.read() },
        automaticSampling: Bool = true
    ) {
        self.reader = reader
        self.modelWasLoadedAtStart = modelWasLoadedAtStart
        self.availableMemorySupported = availableMemorySupported
        queue.sync { takeSample(trigger: .start) }
        if automaticSampling {
            let source = DispatchSource.makeTimerSource(queue: queue)
            source.schedule(
                deadline: .now() + .milliseconds(Self.intervalMilliseconds),
                repeating: .milliseconds(Self.intervalMilliseconds),
                leeway: .milliseconds(5)
            )
            source.setEventHandler { [weak self] in self?.takeSample(trigger: .periodic) }
            timer = source
            source.resume()
        }
    }

    deinit {
        timer?.cancel()
    }

    public func sampleNow() {
        queue.sync {
            guard finishedReport == nil else { return }
            takeSample(trigger: .manual)
        }
    }

    public func finish(memoryWarnings: Int?, lifecycleInterruptions: Int) -> ProcessMemoryReport {
        queue.sync {
            if let finishedReport { return finishedReport }
            timer?.cancel()
            timer = nil
            takeSample(trigger: .end)
            let report = ProcessMemoryReport(
                startedAtUnixSeconds: startDate.timeIntervalSince1970,
                finishedAtUnixSeconds: Date().timeIntervalSince1970,
                monotonicDurationSeconds: elapsed(),
                sampleIntervalMilliseconds: Self.intervalMilliseconds,
                availableMemorySupported: availableMemorySupported,
                modelWasLoadedAtStart: modelWasLoadedAtStart,
                memoryWarnings: memoryWarnings,
                lifecycleInterruptions: lifecycleInterruptions,
                samples: samples,
                samplingFailures: failures
            )
            finishedReport = report
            return report
        }
    }

    private func takeSample(trigger: ProcessMemorySampleTrigger) {
        guard finishedReport == nil else { return }
        switch reader() {
        case .success(let reading):
            samples.append(
                ProcessMemorySample(trigger: trigger, elapsedSeconds: elapsed(), reading: reading))
        case .failure(let error):
            let message = error.localizedDescription
            logger.error("Process memory sampling failed: \(message, privacy: .public)")
            failures.append(
                ProcessMemorySampleFailure(
                    trigger: trigger, elapsedSeconds: elapsed(), message: message))
        }
    }

    private func elapsed() -> Double {
        let duration = startClock.duration(to: .now).components
        return Double(duration.seconds) + Double(duration.attoseconds) / 1e18
    }
}
